File: api/videos.js
import { API_CONFIG } from '../constants/config';
import { apiClient } from './apiClient';

/**
 * Generate Instagram-optimized video using OpenAI
 * WARNING: This is an async operation that takes 30-60 seconds
 * the video content is optimized via prompts for best visual results.
 * @param {string} prompt - Description of video to generate
 * @param {string} contentType - 'reel', 'story', or 'post'
 * @param {string} style - Style: professional, creative, minimal, vibrant, cinematic, natural
 * @returns {Promise<Object>} Response with id, file_url (S3 URL), created_at
 */
/**
 * Check video generation status
 */
async function checkVideoStatus(videoId) {
  const url = `${API_CONFIG.BASE_URL}/content/videos/status/${videoId}`;
  const res = await apiClient.get(url);
  if (!res.ok) {
    throw new Error(`Failed to check video status: ${res.status}`);
  }
  return await res.json();
}

/**
 * Poll for video completion
 */
async function pollVideoCompletion(videoId, maxAttempts = 60, intervalMs = 5000) {
  for (let i = 0; i < maxAttempts; i++) {
    const status = await checkVideoStatus(videoId);
    
    if (status.status === 'completed' && status.file_url) {
      return status;
    }
    
    if (status.status === 'failed') {
      throw new Error(status.error || 'Video generation failed');
    }
    
    // Wait before next poll
    await new Promise(resolve => setTimeout(resolve, intervalMs));
  }
  
  throw new Error('Video generation timed out');
}

export async function generateVideo(prompt, contentType = 'reel', style = null) {
  const url = `${API_CONFIG.BASE_URL}/content/videos/`;

  try {
    // Start video generation (returns immediately with job ID)
    const res = await apiClient.post(url, { 
      prompt,
      content_type: contentType,
      style: style
    });

    if (!res.ok) {
      let errorMessage = `Video generation failed (${res.status})`;
      
      try {
        const errorData = await res.json();
        if (errorData.detail) {
          errorMessage = errorData.detail;
        }
      } catch (parseErr) {
        try {
          const errorText = await res.text();
          if (errorText) {
            errorMessage = errorText;
          }
        } catch (textErr) {
          // Use default error message
        }
      }
      
      console.error("Error response:", errorMessage);
      throw new Error(errorMessage);
    }

    const data = await res.json();
    
    // If status is processing, poll for completion
    if (data.status === 'processing') {
      console.log('Video generation started, polling for completion...');
      const completed = await pollVideoCompletion(data.id);
      return {
        id: completed.id,
        file_url: completed.file_url,
        created_at: data.created_at
      };
    }
    
    return data;

  } catch (err) {
    console.error("ERROR inside generateVideo:", err);
    throw err;
  }
}

/**
 * Fetch all videos for authenticated user
 * @returns {Promise<Array>} Array of video content
 */
export async function fetchAllVideos() {
  const url = `${API_CONFIG.BASE_URL}/content/videos/`;

  try {
    const res = await apiClient.get(url);

    if (!res.ok) {
      throw new Error(`Failed to fetch videos: ${res.status}`);
    }

    return await res.json();

  } catch (err) {
    console.error("ERROR fetching videos:", err);
    throw err;
  }
}

/**
 * Delete a video
 * @param {number} videoId - ID of video to delete
 */
export async function deleteVideo(videoId) {
  const url = `${API_CONFIG.BASE_URL}/content/videos/${videoId}`;

  try {
    const res = await apiClient.delete(url);

    if (!res.ok) {
      throw new Error(`Failed to delete video: ${res.status}`);
    }

    return true;

  } catch (err) {
    console.error("ERROR deleting video:", err);
    throw err;
  }
}

File: backend/Project_Content/ContentApp/routers/videos.py
from typing import Annotated, Optional
from pydantic import BaseModel, Field
from fastapi import APIRouter, Depends, HTTPException, status, BackgroundTasks
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from .auth import get_current_user
from ..services.bvideo import generate_video
from ..firebase_db import GeneratedContentDB, ActivityDB
import asyncio
from concurrent.futures import ThreadPoolExecutor

# Thread pool for running sync video generation in background
_executor = ThreadPoolExecutor(max_workers=2)

router = APIRouter(
    prefix='/content/videos',
    tags=['videos generation']
)

user_dependency = Annotated[dict, Depends(get_current_user)]
security = HTTPBearer(auto_error=False)

async def get_current_user_optional(credentials: Optional[HTTPAuthorizationCredentials] = Depends(security)):
    """Get current user from token if provided, otherwise return None"""
    if credentials is None:
        return None
    try:
        from jose import jwt
        from .auth import SECRET_KEY, ALGORITHM
        token = credentials.credentials
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        user_id: str = payload.get('id')
        username: str = payload.get('sub')
        user_role: str = payload.get('role')
        if username is None or user_id is None:
            return None
        return {'username': username, 'id': user_id, 'user_role': user_role}
    except:
        return None

user_dependency_optional = Annotated[Optional[dict], Depends(get_current_user_optional)]


class RequestVideo(BaseModel):
    prompt: str = Field(min_length=3)
    content_type: str = Field(default='reel', description="Content type: 'reel', 'story', or 'post'")
    style: Optional[str] = Field(default=None, description="Style: professional, creative, minimal, vibrant, cinematic, natural")



@router.get("/", status_code=status.HTTP_200_OK)
async def fetch_all_videos(user: user_dependency):
    """Fetch all video content for the current user"""
    if user is None:
        raise HTTPException(status_code=401, detail='Authentication Failed')
    
    results = GeneratedContentDB.get_by_owner(user["id"], content_type="video")
    return results


@router.get("/{video_id}", status_code=status.HTTP_200_OK)
async def read_video(user: user_dependency, video_id: str):
    """Get a specific video by ID"""
    if user is None:
        raise HTTPException(status_code=401, detail='Authentication Failed')

    video = GeneratedContentDB.get_by_id(video_id)
    
    if not video or video.get('owner_id') != user["id"] or video.get('type') != 'video':
        raise HTTPException(status_code=404, detail="Video not found")

    return video


@router.delete("/{video_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_video(user: user_dependency, video_id: str):
    """Delete a video"""
    if user is None:
        raise HTTPException(status_code=401, detail='Authentication Failed')

    video = GeneratedContentDB.get_by_id(video_id)
    
    if not video or video.get('owner_id') != user["id"]:
        raise HTTPException(status_code=404, detail='Content not found.')
    
    success = GeneratedContentDB.delete(video_id)
    if not success:
        raise HTTPException(status_code=404, detail="Video not found or already deleted.")


def _generate_video_background(video_id: str, prompt: str, content_type: str, style: str, user_id: str = None):
    """Background task to generate video and update the record"""
    try:
        data_url = generate_video(
            prompt=prompt,
            content_type=content_type,
            style=style
        )
        
        # Update the video record with the generated URL
        GeneratedContentDB.update(video_id, {
            'file_url': data_url,
            'status': 'completed'
        })
        
        # Log activity
        if user_id:
            try:
                ActivityDB.create({
                    'user_id': user_id,
                    'type': 'content_generated',
                    'action': 'Generated Video',
                    'description': f'AI generated video: {prompt[:50]}...',
                    'content_type': 'video',
                    'content_id': video_id,
                    'metadata': {
                        'prompt': prompt,
                        'file_url': data_url
                    }
                })
            except:
                pass
                
    except Exception as e:
        # Update status to failed
        GeneratedContentDB.update(video_id, {
            'status': 'failed',
            'error': str(e)
        })


@router.post("/", status_code=status.HTTP_202_ACCEPTED)
async def create_video(request: RequestVideo, background_tasks: BackgroundTasks, user: user_dependency_optional = None):
    """Generate an Instagram-optimized video using AI (async - returns immediately)"""
    
    # Create a pending video record
    content_data = {
        'type': 'video',
        'prompt': request.prompt,
        'file_url': None,
        'status': 'processing',
        'owner_id': user['id'] if user else None
    }
    
    new_video = GeneratedContentDB.create(content_data)
    
    # Start video generation in background
    background_tasks.add_task(
        _generate_video_background,
        new_video['id'],
        request.prompt,
        request.content_type,
        request.style,
        user['id'] if user else None
    )

    return {
        "id": new_video['id'],
        "status": "processing",
        "message": "Video generation started. Poll /content/videos/{id} to check status.",
        "created_at": str(new_video['created_at'])
    }


@router.get("/status/{video_id}", status_code=status.HTTP_200_OK)
async def get_video_status(video_id: str, user: user_dependency_optional = None):
    """Check the status of a video generation job"""
    video = GeneratedContentDB.get_by_id(video_id)
    
    if not video:
        raise HTTPException(status_code=404, detail="Video not found")
    
    # Check ownership if user is authenticated
    if user and video.get('owner_id') and video.get('owner_id') != user['id']:
        raise HTTPException(status_code=404, detail="Video not found")
    
    return {
        "id": video_id,
        "status": video.get('status', 'unknown'),
        "file_url": video.get('file_url'),
        "error": video.get('error')
    }





